torch/onnx/_internal/exporter/_core.py

```diff
@@ -99,17 +99,18 @@ def __init__(self, tensor: torch.Tensor, name: str | None = None):
 
     def __array__(self, dtype: Any = None) -> np.ndarray:
         # numpy() calls __array__ in ir.Tensor
+        self.raw: torch.Tensor
         if self.dtype == ir.DataType.BFLOAT16:
-            return self.raw.view(torch.uint16).__array__(dtype)
+            return self.raw.view(torch.uint16).numpy(force=True).__array__(dtype)
```

Comment:
Maybe for a future change, is there a way to keep torch.Tensor?

Reply:
The tensor can be accessed via .raw, or have its content copied via dlpack. The array magic will always be numpy array. Does that sound good to you?

Reply:
Yes, I should have asked my second question. What is used when the model is serialized and for constant folding?

Reply:
For serialization, tobytes is called. For constant folding, numpy is used because we use the reference implementation

Reply:
I see. ``tensor = self.raw.detach().cpu().contiguous()`` should probably be written with contiguous first ``tensor = self.raw.contiguous().detach().cpu()``. It is not necessarily a cheap. For the reference evaluator, maybe we can update it to support torch Tensor (through the array API) so that the cost if cheaper as well. That will wait for the time being.

Reply:
According to https://pytorch.org/docs/stable/generated/torch.Tensor.numpy.html  this is the same as t.detach().cpu().resolve_conj().resolve_neg().numpy() so contiguous is not called?

Reply:
I'm saying it should happen on cuda not cpu.

Reply:
resolve conj and resolve neg will materialize some flags I think. I am speculating torch didn’t want to change the original tensor.

Reply:
But judging from the nature of the tensors we are handling (weights) they will most likely be no-ops.

Reply:
Not sure I follow the above discussion ... I don't see any call to detach(), is that not required? Anyway: with regards to perf implications, we should evaluate and decide and that is a separate issue for the future.

Reply:
The force=True option will call everything above including detach